docs/ARCOptimization.md

```diff
@@ -0,0 +1,947 @@
+optimization on cases, one has to be very careful about ensuring that
+one only ignores reference count operations on values that are able to
+be proved to be that specific case.
+
+*TODO*: This section needs to be filled out more.
+
+## Copy-On-Write Considerations
+
+The copy-on-write capabilities of some data structures, such as Array
+and Set, are efficiently implemented via Builtin.isUnique calls which
+lower directly to is_unique instructions in SIL.
+
+The is_unique instruction takes the address of a reference, and although
+it does not actually change the reference, the reference must appear
+mutable to the optimizer. This forces the optimizer to preserve a retain
+distinct from what's required to maintain lifetime for any of the
+reference's source-level copies, because the called function is allowed
+to replace the reference, thereby releasing the referent. Consider the
+following sequence of rules:
+
+(1) An operation taking the address of a variable is allowed to replace
+    the reference held by that variable. The fact that is_unique will
+    not actually replace it is opaque to the optimizer.
+(2) If the refcount is 1 when the reference is replaced, the referent is
+    deallocated.
+(3) A different source-level variable pointing at the same referent must
+    not be changed/invalidated by such a call.
+(4) If such a variable exists, the compiler must guarantee the refcount
+    is \> 1 going into the call.
+
+With the is_unique instruction, the variable whose reference is being
+checked for uniqueness appears mutable at the level of an individual SIL
+instruction. After IRGen, is_unique instructions are expanded into
+runtime calls that no longer take the address of the variable.
+Consequently, LLVM-level ARC optimization must be more conservative. It
+must not remove retain/release pairs of this form:
+
+```sil
+retain X
+retain X
+_swift_isUniquelyReferenced(X)
+release X
+release X
+```
+
+To prevent removal of the apparently redundant inner retain/release
+pair, the LLVM ARC optimizer should model `_swift_isUniquelyReferenced`
+as a function that may release X, use X, and exit the program (the
+subsequent release instruction does not prove safety).
+
+### `is_unique` instruction
+
+As explained above, the SIL-level is_unique instruction enforces the
+semantics of uniqueness checks in the presence of ARC optimization. The
+kind of reference count checking that is_unique performs depends on the
+argument type:
+
+> -   Native object types are directly checked by reading the strong
+>     reference count: (Builtin.NativeObject, known native class
+>     reference)
+> -   Objective-C object types require an additional check that the
+>     dynamic object type uses native Swift reference counting: (unknown
+>     class reference, class existential)
+> -   Bridged object types allow the dynamic object type check to be
+>     bypassed based on the pointer encoding: (Builtin.BridgeObject)
```

Comment:
Maybe this was accidentally marked as a quote? Seems a bit odd... It seems like it is stating the rules, and not quoting rules stated elsewhere.